ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl/Pure.hs

```diff
@@ -0,0 +1,321 @@
+    -- * Mempool
+    pureGetSnapshotFor
+  , pureRemoveTxs
+  , pureSyncWithLedger
+  , pureTryAddTxs
+  , runRemoveTxs
+  , runSyncWithLedger
+  , runTryAddTxs
+    -- * MempoolSnapshot
+  , implSnapshotFromIS
+  ) where
+
+import           Control.Exception (assert)
+import           Data.Maybe (isJust, isNothing)
+import qualified Data.Set as Set
+import           Data.Word (Word32)
+
+import           Control.Tracer
+
+import           Ouroboros.Consensus.HeaderValidation
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.Ledger.SupportsMempool
+import           Ouroboros.Consensus.Mempool.API
+import           Ouroboros.Consensus.Mempool.Impl.Types
+import           Ouroboros.Consensus.Mempool.TxSeq (TicketNo, TxTicket (..),
+                     zeroTicketNo)
+import qualified Ouroboros.Consensus.Mempool.TxSeq as TxSeq
+import           Ouroboros.Consensus.Util.IOLike
+
+
+{-------------------------------------------------------------------------------
+  Mempool Implementation
+-------------------------------------------------------------------------------}
+
+-- | Data constructor containing the control flow for 'tryAddTxs'. Can be
+-- interpreted in an IOLike monad with 'runTryAddTxs'.
+data TryAddTxs blk =
+  Done [GenTx blk]
+  | Trace (TraceEventMempool blk) (TryAddTxs blk)
+  | WriteTx (InternalState blk) (TryAddTxs blk)
+  | Processed (MempoolAddTxResult blk) (TryAddTxs blk)
+
+-- | Add a bunch of transactions (oldest to newest) by interpreting a 'TryAddTxs'
+-- from 'pureTryAddTxs'.
+--
+-- This function returns two lists: the transactions that were added or
+-- rejected, and the transactions that could not yet be added, because the
+-- Mempool capacity was reached. See 'addTxs' for a function that blocks in
+-- case the Mempool capacity is reached.
+--
+-- Transactions are added one by one, updating the Mempool each time one was
+-- added successfully.
+--
+-- This function does not sync the Mempool contents with the ledger state in
+-- case the latter changes, it relies on the background thread to do that.
+--
+-- POSTCONDITON:
+-- > (processed, toProcess) <- runTryAddTxs ... $ pureTryAddTxs ... txs
+-- > map fst processed ++ toProcess == txs
+runTryAddTxs
+  :: forall m blk. IOLike m
+  => StrictTVar m (InternalState blk)
+  -- ^ The TVar containing the internal state of the mempool
+  -> Tracer m (TraceEventMempool blk)
+  -- ^ The tracer from the mempool environment
+  -> TryAddTxs blk
+  -- ^ The value to interpret.
+  -> STM m ( [MempoolAddTxResult blk]
+           , [GenTx blk])
+runTryAddTxs stateVar trcr = go []
+  where go processed (Done remaining) =
+          return (reverse processed, remaining)
+        go processed (Trace tem tat') = do
+          return $ traceWith trcr tem
+          go processed tat'
+        go processed (WriteTx is tat') = do
+          writeTVar stateVar is
+          -- TODO: are intermediate states even needed? or could we stack them
```

Comment:
Since the mini protocols for multiple peers are running concurrently, I anticipate that writing the intermediate states ASAP is in some sense the most "fair"? This might deserve some discussion on a sync call.

Reply:
As this whole operation is done atomically, I don't think this is actually relevant and we can just write once.

Reply:
For now I will leave it as is, probably will change it to be only one write based on previous discussions.

Reply:
I commented elsewhere just a minute ago, and now I realized it was relevant to this thread.

Summary: the "whole operation is done atomically" is new behavior; it previous was not atomic: there was one `STM` transaction per tx. (Note that `atomically` was used once per call of the recursive function named `go` in the old code.)

Reply:
Indeed you are right, answered in the other thread. Closing this one.